packages/flutter/lib/src/services/deferred_component.dart

```diff
@@ -0,0 +1,157 @@
+  installed,
+  /// An installation request has been asked to be canceled.
+  cancelling,
+  /// An installation request has been canceled.
+  canceled,
+  /// An installation request has failed.
+  failed,
+  /// The deferred component has not been requested nor installed or is unavailable.
+  unknown,
+}
+
+/// Manages the installation and loading of deferred component modules.
+///
+/// Deferred components allow Flutter applications to download precompiled AOT
+/// dart code and assets at runtime, reducing the install size of apps and
+/// avoiding installing unnessecary code/assets on end user devices. Common
+/// use cases include deferring installation of advanced or infrequently
+/// used features and limiting locale specific features to users of matching
+/// locales.
+class DeferredComponent {
+  // This class is not meant to be instantiated or extended; this constructor
+  // prevents instantiation and extension.
+  // ignore: unused_element
+  DeferredComponent._();
+
+  // TODO(garyq): We should eventually expand this to install modules by loadingUnitId
+  // as well as moduleName, but currently, loadingUnitId is opaque to the dart code
+  // so this is not possible. The API has been left flexible to allow adding
+  // loadingUnitId as a parameter.
+
+  /// Requests that an assets-only deferred component identified by the [moduleName]
+  /// be downloaded and installed.
+  ///
+  /// This method returns a Future<void> that will complete when the feature is
+  /// installed and any assets are ready to used. When an error occurs, the
+  /// future will complete an error.
+  ///
+  /// This method should be used for asset-only deferred components. Deferred components
+  /// containing dart code should call `loadLibrary()` on a deferred imported
+  /// library's prefix to ensure that the dart code is properly loaded as
+  /// `loadLibrary()` will provide the loading unit id needed for the dart
+  /// library loading process. For example:
+  ///
+  /// ```dart
+  /// import 'split_module.dart' deferred as SplitModule;
+  /// ...
+  /// SplitModule.loadLibrary();
+  /// ```
+  ///
+  /// This method will not load associated dart libraries contained in the dynamic
+  /// feature module, though it will download the files necessary and subsequent
+  /// calls to `loadLibrary()` to load will complete faster.
+  static Future<void> installDeferredComponent({@required String? moduleName}) async {
+    await SystemChannels.deferredComponent.invokeMethod<void>(
+      'installDeferredComponent',
+      <String, dynamic>{ 'loadingUnitId': -1, 'moduleName': moduleName },
+    );
+  }
+  /// Gets the current installation state of the deferred component identified by the
+  /// [moduleName].
+  ///
+  /// This method returns a string that represents the state. Depending on
+  /// the implementation, this string may vary, but the default Google Play
+  /// Store implementation retuns a state in the [DeferredComponentInstallState] enum.
```

Comment:
This is a strange API.

Reply:
I have signed up for API design review on Tuesday, hopefully we can get some opinions on this API, along with the rest of the deferred components system.